### PR TITLE
EXTREMELY CURSED PROOF OF CONCEPT: Standardized Singletons and SLists

### PR DESCRIPTION
**Pull request by Kapu1178**
## About The Pull Request
Changed Highlander Datum to a singleton, requested by arm.
I was having too much fun and created a monster.

So, what the fuck are you looking at? This is a proof-of-concept for standardized singletons, which are managed using macros.

`/singleton` is an abstract type intended to be used like a datum. Highlander example included as a very basic display.

`/singleton/datum` is a singleton based off an existing datum type which you feed it in it's constructor.

`/singleton/slist` is a list singleton, which on mutation, creates a new singleton based on a copy instead of overwriting the original. It's essentially a cached list, where the original isn't mutated to maintain it's state. Why is it "slist"? Because Byond freaks out when you try to call procs on a class that has `/list` in the path, despite every other part of the code working fine.

Singleton Lists are the most developed part of this concept as of right now. You can "modify" singleton lists using `SINGLETON_LIST_MUTATE()`, `SINGLETON_LIST_ADD()`, or `SINGLETON_LIST_REMOVE()`. Byond comparison operators work as normal thanks to the power of OPERATOR OVERLOADING. You can also access the singleton list's internal list using `slist[key]`, just like a normal list!

### Why does this PR exist?
Thanks for asking! I wrote this code at 1 am, and I don't even know if the idea itself is worth persuing. Github is the closest thing to a TG-related forum im going to touch, and discussion on something *this big* shouldn't be relegated to the information blackhole that is discord. 

TL;DR, feedback/thoughts on the concept as a whole, as well as suggestions for how to make this less (or more?) cursed.

### CODE REVIEW

**Comment by tralezab** [comment]
### I LIKE THE /SINGLETON IDEA
wednesday-good

File: code/modules/admin/verbs/highlander_datum.dm [hunk 1]
```diff
@@ -1,13 +1,10 @@
-
-GLOBAL_DATUM(highlander_controller, /datum/highlander_controller)
```
Comment (LemonInTheDark) [inline code/modules/admin/verbs/highlander_datum.dm:2]: I do not understand why you would want to do this

File: code/__DEFINES/singletons.dm [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#define SINGTYPE_ABSTRACT "abstract"
+#define SINGTYPE_COMPLEX "complex"
+
+///Find or create an abstract singleton of the desired type.
+#define SINGLETON(_datatype) _singleton(_datatype)
+///Return a single or null using the given search arguments.
+#define FIND_SINGLETON(_datatype, args...) (islist(global.singleton_repo[_datatype]) ? (global.singleton_repo[_datatype][list(##args).Join()]) : (global.singleton_repo[_datatype]))
```
Comment (LemonInTheDark) [inline code/__DEFINES/singletons.dm:7]: why are you using raw globals? doesn't that make these unvvable without sdql?

File: code/datums/singletons/singleton_lists.dm [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+	return data[value]
+
+/singleton/slist/proc/operator[]=(key, value)
+	CRASH("Unsupported operation! Use SINGLETON_LIST_MUTATE()!")
+
+/singleton/slist/proc/operator|(list/value)
+	return _singleton_list(data.Copy()|value).get()
+
+/singleton/slist/proc/operator&(list/value)
+	return _singleton_list(data.Copy()&value).get()
```
Comment (LemonInTheDark) [inline code/datums/singletons/singleton_lists.dm:38]: I really do not like this

File: code/datums/singletons/singleton_lists.dm [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+	global.singleton_repo[/list][key] = src
+
+	data = _data
+
+/singleton/slist/proc/get()
+	return data.Copy()
+
+/singleton/slist/proc/Copy()
+	return get()
```
Comment (LemonInTheDark) [inline code/datums/singletons/singleton_lists.dm:26]: but why tho

File: code/datums/singletons/singleton_lists.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/datums/singletons/singleton_lists.dm:38]: also use spaces please

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]
I don't really understand why you'd want to use standard singletons over globals, and the use of singleton lists seems like just advanced key'd lists with less control.
Also the operator overloads make my mightily depressed

**Comment by Mothblocks** [comment]
I also extremely do not see the benefit in this and foresee it making testability harder, since it would now not be as straightforward to instantiate a brand new copy for isolated testing purposes, which is an issue that happens frequently with similar patterns.

**Comment by ike709** [comment]
~~Please don't utilize operator overloading because then I would need to implement it in OD to make unmodified TG compile~~
